### DISCUSSION TITLE
Performance metrics

### DISCUSSION THREAD

**Discussion by steveodhiambo** [tendermint/tendermint#7956] [Q&A]
Hey, I would like to do some performance measurements on tendermint  BFT protocol in terms of latency, throughput. Any pointers on how to go about this?

**Comment by creachadair** [comment]
Hi @steveodhiambo. I unfortunately do not have a quick or simple answer here. The performance of Tendermint as a system depends on the number, size, distribution, and workload of the nodes in the network. To a first approximation, a meaningful performance test requires setting up a network to simulate the configuration you expect, and generating some work for it to do.

Even if you're mostly interested in "overhead", where application performance isn't a factor, the costs of consensus are affected by the number and size of transactions submitted for consideration, and the number of peers (nodes in the network) and clients submitting transactions. It's fairly easy to spin up a few nodes locally, but such a configuration won't give much signal about performance in more realistic conditions. A more effective test might start by creating a few cooperating nodes on different networks (e.g., different cloud zones or hosting providers).

The Tendermint core does expose a number of Prometheus metrics that may help with gathering the data you need (and if you find there are metrics you want that are missing, please let us know!) Moreover, you can enable pprof on a running node (with `--rpc.pprof-laddr` or in the config file) if you need to capture CPU or memory profiles for individual nodes. In general, though, the behaviour of a single node is not that informative about end-to-end performance. Profiles are usually more interesting as a way of diagnosing specific issues ("this node is thrashing").

So, in short: The best answer I can offer is to try to simulate, as best as conditions allow, the environment you'd expect to run in production. Networking is a big factor in BFT consensus performance, so that's the biggest factor to grapple with.

I hope that will at least help you get started.

**Reply by steveodhiambo** [reply]
Thanks
